Sources/Views/MessageLabel.swift

```diff
@@ -22,415 +22,477 @@
-
+        let textOffset = CGPoint(x: textInsets.left, y: textInsets.right) // <- bug?
+        
         location.x -= textOffset.x
         location.y -= textOffset.y
-
+        
         let glyphIndex = layoutManager.glyphIndex(for: location, in: textContainer)
-
+        
         let lineRect = layoutManager.lineFragmentUsedRect(forGlyphAt: glyphIndex, effectiveRange: nil)
-
-        if lineRect.contains(location) {
-            return layoutManager.characterIndexForGlyph(at: glyphIndex)
-        } else {
-            return nil
-        }
-
+        
+        let rectContainsLocation = lineRect.contains(location)
+        
+        return rectContainsLocation ? layoutManager.characterIndexForGlyph(at: glyphIndex) : nil
+        
     }
-
+    
     private func setupGestureRecognizers() {
-
+        
         let tapGesture = UITapGestureRecognizer(target: self, action: #selector(handleGesture(_:)))
         addGestureRecognizer(tapGesture)
         tapGesture.delegate = self
-
+        
         let longPressGesture = UILongPressGestureRecognizer(target: self, action: #selector(handleGesture(_:)))
         addGestureRecognizer(longPressGesture)
         tapGesture.delegate = self
-
+        
         isUserInteractionEnabled = true
     }
-
-  @objc func handleGesture(_ gesture: UIGestureRecognizer) {
-
+    
+    @objc
+    func handleGesture(_ gesture: UIGestureRecognizer) {
+        
         let touchLocation = gesture.location(ofTouch: 0, in: self)
         guard let index = stringIndex(at: touchLocation) else { return }
-
-        for (detectorType, ranges) in rangesForDetectors {
-            for (nsRange, value) in ranges {
-                guard let range = Range(nsRange) else { return }
-                if range.contains(index) {
-                    handleGesture(for: detectorType, value: value)
-                }
+        
+        var detectorRangeMap: [DetectorType: [NSRange]] = [:]
+        
+        for detector in enabledDetectors where !enabledDetectors.isEmpty {
+            detectorRangeMap[detector] = detectedRanges(for: detector)
+        }
+        
+        for (detectorType, ranges) in detectorRangeMap {
+            for nsRange in ranges {
+                guard let range = Range(nsRange), range.contains(index) else { continue }
+                let results = detectorResults(for: detectorType)
+                guard !results.isEmpty, let value = results[nsRange] else { continue }
+                forwardToDelegateMethod(for: detectorType, value: value)
             }
         }
     }
-
-    private func handleGesture(for detectorType: DetectorType, value: Any?) {
-
-        switch detectorType {
+    
+    private func forwardToDelegateMethod(for detector: DetectorType, value: Any?) {
+        switch detector {
         case .address:
             guard let addressComponents = value as? [String: String] else { return }
             handleAddress(addressComponents)
```

Comment:
This is broken because `value: Any?` is a `[NSTextCheckingKey : String]?`. I'm not sure how this cast worked in the old code to be honest. I know `NSTextCheckingKey` has a `RawValue` of `String`. 

Maybe the old code was so bad that I'm A) missing something, B) got away with some sort of lossy conversion. 

Point is I have to either break the API and pass in `[NSTextCheckingKey: String]` or somehow get to `[String: String]` again. I don't mind making a change here as opposed to mapping the collection